File: android/app/src/main/java/com/agentcontrolcenter/app/runtime/agent/AgentManager.kt
package com.agentcontrolcenter.app.runtime.agent

import com.agentcontrolcenter.app.agent.model.Agent
import com.agentcontrolcenter.app.agent.model.AgentCapability
import com.agentcontrolcenter.app.agent.model.AgentStatus
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Agent 管理器 — Agent Control Center 的核心调度中心。
 *
 * 职责：
 * - 维护所有已注册 Agent 的生命周期
 * - 提供 Agent 状态的响应式流
 * - 协调 Transport 层与 Agent 实例的绑定
 *
 * 设计理念：Agent 是第一公民，不再以 Chat 为中心。
 * User → Agent → Chat/Task/Workflow/Tool/Memory
 *
 * 线程安全：@Singleton 会被多协程并发访问，所有 StateFlow 写入均通过
 * `update { }` 原子完成，避免「读-改-写」竞争丢失并发更新。
 *
 * Phase 4.3: 注入 [AgentRegistry]，registerAgent/unregisterAgent 同步到
 * Registry 以支持按 Capability 查询。getAgentsByCapability 委托给 Registry。
 */
@Singleton
class AgentManager @Inject constructor(
    private val registry: AgentRegistry
) {

    private val _agents = MutableStateFlow<List<Agent>>(emptyList())
    val agents: StateFlow<List<Agent>> = _agents.asStateFlow()

    private val _activeAgent = MutableStateFlow<Agent?>(null)
    val activeAgent: StateFlow<Agent?> = _activeAgent.asStateFlow()

    fun registerAgent(agent: Agent) {
        _agents.update { current ->
            val index = current.indexOfFirst { it.id == agent.id }
            if (index >= 0) current.toMutableList().also { it[index] = agent } else current + agent
        }
        // Phase 4.3: 同步到 Registry，支持按 Capability 查询
        registry.register(agent)
    }

    fun unregisterAgent(agentId: String) {
        _agents.update { it.filter { agent -> agent.id != agentId } }
        if (_activeAgent.value?.id == agentId) {
            _activeAgent.value = null
        }
        // Phase 4.3: 同步到 Registry
        registry.unregister(agentId)
    }

    fun setActiveAgent(agentId: String) {
        _activeAgent.value = _agents.value.find { it.id == agentId }
    }

    fun updateAgentStatus(agentId: String, status: AgentStatus) {
        var updatedAgent: Agent? = null
        _agents.update { current ->
            val index = current.indexOfFirst { it.id == agentId }
            if (index >= 0) {
                val updated = current[index].copy(status = status)
                updatedAgent = updated
                current.toMutableList().also { it[index] = updated }
            } else {
                current
            }
        }

        // Phase 4.3: 同步状态变更到 Registry
        updatedAgent?.let { registry.register(it) }
    }

    fun getAgent(agentId: String): Agent? = _agents.value.find { it.id == agentId }

    /**
     * Phase 4.3: 按 [capability] 过滤 Agent，委托给 [AgentRegistry]。
     * 若 Registry 中无数据（未注册），回退到内存列表的 capabilities 过滤。
     */
    fun getAgentsByCapability(capability: AgentCapability): List<Agent> {
        val fromRegistry = registry.getByCapability(capability)
        if (fromRegistry.isNotEmpty()) return fromRegistry
        // 回退：直接过滤内存列表
        return _agents.value.filter { it.capabilities.contains(capability) }
    }
}

File: ios/AgentControlCenter/Runtime/AgentManager.swift
import Foundation
import Observation

// MARK: - AgentManager
// 对应 Android AgentManager + AgentRegistry

/// Agent 管理器 — Agent Control Center 的核心调度中心。
///
/// 职责：
/// - 维护所有已注册 Agent 的生命周期
/// - 提供 Agent 状态的响应式流
/// - 协调 Transport 层与 Agent 实例的绑定
/// - 按 Capability 查询 Agent
///
/// `@MainActor` 隔离保证 `agents` / `activeAgent` 等响应式状态的读写
/// 均在主线程进行，避免 SwiftUI 视图读取时发生数据竞争。
@MainActor
@Observable
final class AgentManager {

    /// 所有已注册的 Agent
    private(set) var agents: [Agent] = []

    /// 当前活跃的 Agent
    private(set) var activeAgent: Agent?

    /// Agent 能力索引 (capability -> agentId 集合)
    private var capabilityIndex: [AgentCapability: Set<String>] = [:]

    // MARK: - Registration

    /// 注册或更新 Agent
    func register(_ agent: Agent) {
        if let index = agents.firstIndex(where: { $0.id == agent.id }) {
            agents[index] = agent
        } else {
            agents.append(agent)
        }
        // 更新能力索引
        for cap in agent.capabilities {
            capabilityIndex[cap, default: []].insert(agent.id)
        }
    }

    /// 注销 Agent
    func unregister(agentId: String) {
        // 从能力索引中移除
        for (cap, ids) in capabilityIndex {
            var updated = ids
            updated.remove(agentId)
            if updated.isEmpty {
                capabilityIndex.removeValue(forKey: cap)
            } else {
                capabilityIndex[cap] = updated
            }
        }
        agents.removeAll { $0.id == agentId }
        if activeAgent?.id == agentId {
            activeAgent = nil
        }
    }

    /// 设置活跃 Agent
    func setActive(agentId: String) {
        activeAgent = agents.first { $0.id == agentId }
    }

    /// 更新 Agent 状态
    func updateStatus(agentId: String, status: AgentStatus) {
        guard let index = agents.firstIndex(where: { $0.id == agentId }) else { return }
        var updated = agents[index]
        updated.status = status
        agents[index] = updated
    }

    // MARK: - Query

    /// 获取 Agent
    func getAgent(_ agentId: String) -> Agent? {
        agents.first { $0.id == agentId }
    }

    /// 按 Capability 查询 Agent
    func getAgentsByCapability(_ capability: AgentCapability) -> [Agent] {
        let agentIds = capabilityIndex[capability] ?? []
        return agents.filter { agentIds.contains($0.id) }
    }

    /// 获取所有在线 Agent
    var onlineAgents: [Agent] {
        agents.filter { $0.status == .online }
    }
}
